Add Kotlin DSL for building *Config and *Registry classes

GreyTeardrop opened the pull request:
## Overview

This proposal adds a DSL for building `BulkheadConfig`, `CircuitBreakerConfig`, `RateLimiterConfig`, `RetryConfig` and `TimeLimiterConfig`:

```kotlin
val circuitBreakerConfig = CircuitBreakerConfig {
    failureRateThreshold(50)
    waitDurationInOpenState(Duration.ofSeconds(30))
}

// instead of
val circuitBreakerConfig = CircuitBreakerConfig.custom()
    .failureRateThreshold(50)
    .waitDurationInOpenState(Duration.ofSeconds(30))
    .build()
```

In addition to a bit nicer syntax, this DSL also opens up a possibility for conditions in the configuration builders:
```kotlin
val circuitBreakerConfig = CircuitBreakerConfig {
    failureRateThreshold(50)
    waitDurationInOpenState(Duration.ofSeconds(30))
    if (limitSlowCalls) {
        slowCallDurationThreshold(Duration.ofSeconds(60))
        slowCallRateThreshold(20)
    }
}
```

**Open question**: all DSL methods are inline so it's hard to directly measure their code coverage. At the moment these builders are used in existing `resilience4j-kotlin` tests but no new tests were added.

RobWin commented:
Thanks. Could you also please create an issue

RobWin commented:
DSL looks good. Good idea. Thank you.

RobWin commented:
We also introduced builders for Registries.
Would you also like to add a DSL for Registries as well?

GreyTeardrop commented:
> Thanks. Could you also please create an issue

I've created #964 and copied the description from this PR there. Is that ok, or should the issue have something else?

GreyTeardrop commented:
> We also introduced builders for Registries.
> Would you like to add a DSL for Registries as well?

Totally makes sense, I'll work on adding DSL for Registries as well.

RobWin commented:
> I've created #964 and copied the description from this PR there. Is that ok, or should the issue have something else?

That's fine. Thanks

GreyTeardrop commented:
@RobWin I've added DSLs for all `Registry.Builder`s. For `BulkheadRegistry.Builder` I've also added several method overloads to make it more "DSL-y":

```kotlin
val bulkheadRegistry = BulkheadRegistry {
    withBulkheadConfig {
        maxConcurrentCalls(2)
        maxWaitDuration(Duration.ZERO)
    }
    addBulkheadConfig("sharedConfig1") {
        maxConcurrentCalls(3)
        maxWaitDuration(Duration.ofSecond(1))
    }
    withTags("tag1" to "value1", "tag2" to "value2")
}
```

If you feel these make sense I could add them to the rest of Registry DSLs.

RobWin commented:
Yes, I like it.

GreyTeardrop commented:
I've updated PR with dedicated `with*Config` and `add*Config` DSLs for all registry builders.

RobWin commented:
PR looks fine to me

GreyTeardrop commented:
I've just pushed an update for some of KDocs, they had wrong copy-pasted code examples, now all examples should be valid.

RobWin commented:
Thank you

RobWin commented:
Damn I forgot that all source files need the Apache 2.0 Lisence Header. Could you please add them and create another PR?

GreyTeardrop commented:
@RobWin Sure, I can do that now. Should there also be an issue for that or just a PR?

RobWin commented:
Just a PR :) is missed it in my review

GreyTeardrop commented:
@RobWin Does this one look ok? https://github.com/resilience4j/resilience4j/pull/976
